transaction.go

```diff
@@ -23,12 +23,121 @@ import "C"
+	// ErrNewAuthtokReqd indicates a new authentication token required. This is
+	// normally returned if the machine security policies require that the
+	// password should be changed because the password is nil or it has aged.
+	ErrNewAuthtokReqd StatusError = C.PAM_NEW_AUTHTOK_REQD
+	// ErrAcctExpired indicates that an user account has expired.
+	ErrAcctExpired StatusError = C.PAM_ACCT_EXPIRED
+	// ErrSession indicates a can not make/remove an entry for the
+	// specified session.
+	ErrSession StatusError = C.PAM_SESSION_ERR
+	// ErrCredUnavail indicates that an underlying authentication service can not
+	// retrieve user credentials.
+	ErrCredUnavail StatusError = C.PAM_CRED_UNAVAIL
+	// ErrCredExpired indicates that an user credentials expired.
+	ErrCredExpired StatusError = C.PAM_CRED_EXPIRED
+	// ErrCred indicates a failure setting user credentials.
+	ErrCred StatusError = C.PAM_CRED_ERR
+	// ErrNoModuleData indicates a no module specific data is present.
+	ErrNoModuleData StatusError = C.PAM_NO_MODULE_DATA
+	// ErrConv indicates a conversation error.
+	ErrConv StatusError = C.PAM_CONV_ERR
+	// ErrAuthtokErr indicates an authentication token manipulation error.
+	ErrAuthtok StatusError = C.PAM_AUTHTOK_ERR
+	// ErrAuthtokRecoveryErr indicates an authentication information cannot
+	// be recovered.
+	ErrAuthtokRecovery StatusError = C.PAM_AUTHTOK_RECOVERY_ERR
+	// ErrAuthtokLockBusy indicates am authentication token lock busy.
+	ErrAuthtokLockBusy StatusError = C.PAM_AUTHTOK_LOCK_BUSY
+	// ErrAuthtokDisableAging indicates an authentication token aging disabled.
+	ErrAuthtokDisableAging StatusError = C.PAM_AUTHTOK_DISABLE_AGING
+	// ErrTryAgain indicates a preliminary check by password service.
+	ErrTryAgain StatusError = C.PAM_TRY_AGAIN
+	// ErrIgnore indicates to ignore underlying account module regardless of
+	// whether the control flag is required, optional, or sufficient.
+	ErrIgnore StatusError = C.PAM_IGNORE
+	// ErrAbort indicates a critical error (module fail now request).
+	ErrAbort StatusError = C.PAM_ABORT
+	// ErrAuthtokExpired indicates an user's authentication token has expired.
+	ErrAuthtokExpired StatusError = C.PAM_AUTHTOK_EXPIRED
+	// ErrModuleUnknown indicates a module is not known.
+	ErrModuleUnknown StatusError = C.PAM_MODULE_UNKNOWN
+	// ErrBadItem indicates a bad item passed to pam_*_item().
+	ErrBadItem StatusError = C.PAM_BAD_ITEM
+	// ErrConvAgain indicates a conversation function is event driven and data
+	// is not available yet.
+	ErrConvAgain StatusError = C.PAM_CONV_AGAIN
+	// ErrIncomplete indicates to please call this function again to complete
+	// authentication stack. Before calling again, verify that conversation
+	// is completed.
+	ErrIncomplete StatusError = C.PAM_INCOMPLETE
+)
+
+// String returns the string for the given Status, allowing us to implement the
+// [fmt.Stringer] interface.
+func (status StatusError) String() string {
+	return C.GoString(C.pam_strerror(nil, C.int(status)))
```

Comment:
Passing nil to `pam_strerror` doesn't seem right. Have you looked at the pam source to see why that is there? Does `pam_strerror` add extra context if the pam handle is not NULL?

Maybe the error type could be implemented similar to `syscall.Errors`:

https://cs.opensource.google/go/go/+/refs/tags/go1.21.4:src/syscall/syscall_unix.go;l=92-116
https://cs.opensource.google/go/go/+/refs/tags/go1.21.4:src/syscall/zerrors_linux_amd64.go

In this scenario the map in the test case would be part of the code and the strings would be returned from the map instead of calling into `pam_strerror`. I also think that test is probably not needed in either case. Right now I feel like it answers the question: Does `pam_strerror` return this specific string? If pam changes the wording of an error (admittedly unlikely) we would get a test failure here. If we do keep the `pam_strerror` call I think a test to verify that we get any string back (and it doesn't crash) is good. What do you think?

Another option would be to generate the error strings with `pam_strerror` when the error occurs rather than deferring to when `StatusError.Error()` is called.

Reply:
So, yes... I did check the definition and in all the relevant ones it's just ignored:
- https://github.com/linux-pam/linux-pam/blob/8d082da1bc993b5b061ae81a9743891328e04ce6/libpam/pam_strerror.c#L36-L106
 - https://github.com/freebsd/freebsd-src/blob/d565784a7ebaa59e26febdcfd4a60329786ea5f5/contrib/openpam/lib/libpam/pam_strerror.c#L53-L75

So, duplicating it may make sense, but really there's no case where this could go wrong right now.

Reply:
OK, sounds good.